packages/db/tests/collection.test.ts

```diff
@@ -1508,4 +1509,64 @@ describe(`Collection with schema validation`, () => {
     expect(collection.syncedData.size).toBe(0)
     expect(collection.syncedMetadata.size).toBe(0)
   })
+  it(`open sync transaction doesn't get applied when optimistic mutation is resolved`, async () => {
+    type Row = { id: number; name: string }
+
+    const collection = createCollection(
+      mockSyncCollectionOptionsNoInitialState<Row>({
+        id: `repro-truncate-open-transaction`,
+        getKey: (r) => r.id,
+      })
+    )
+    const preloadPromise = collection.preload()
+
+    collection.utils.begin()
+    collection.utils.write({ type: `insert`, value: { id: 1, name: `one` } })
+    collection.utils.write({ type: `insert`, value: { id: 2, name: `two` } })
+    collection.utils.commit()
+    collection.utils.markReady()
+
+    await preloadPromise
+
+    expect(collection.state.size).toBe(2)
+
+    // start a transaction with a truncate, but don't commit it
+    collection.utils.begin()
+    collection.utils.truncate()
+
+    // expect the state to still be the same
+    expect(collection.state.size).toBe(2)
+
+    // we now do a local optimistic insert
+    collection.insert({ id: 3, name: `three` })
+
+    // we should immediately see the optimistic state
+    expect(collection.state.size).toBe(3)
+    expect(collection.state.get(3)?.name).toBe(`three`)
+
+    // resolve the sync, this is a bit of an anti-pattern here as we haven't
+    // actually seen the insert from the server yet... but this could easily happen in
+    // a misconfigured setup.
+    // you would see the same effect if the local mutation was rejected and rolled back.
+    collection.utils.resolveSync()
+    await new Promise((resolve) => setTimeout(resolve, 0))
+
+    // we should now be back to the original state
+    expect(collection.state.size).toBe(2)
```

Comment:
We no longer see the optimistic update because we called `resolveSync` and that assumes that the write has already made its way back to the client?
And since that's not the case here, that optimistic update is gone. Is that right?
Still not completely sure why we can't just let the write sync back before calling `resolveSync`.